### PR TITLE
user_role_workflow_manager bug fixed

### PR DESCRIPTION
**Pull request by ajithandrewj**
## Description
The below issue is fixed 
1 Security Issue: Plaintext Password Storage in User and Role Workflow Manager

## Type of Change
- [ yes ] Bug fix
- [ no ] New feature
- [ no ] Breaking change
- [ no ] Documentation update

## Checklist
- [ yes ] My code follows the style guidelines of this project
- [ yes ] I have performed a self-review of my own code
- [ yes ] I have commented my code, particularly in hard-to-understand areas
- [ no ] I have made corresponding changes to the documentation
- [ yes ] My changes generate no new warnings
- [ yes ] I have added tests that prove my fix is effective or that my feature works
- [ yes ] New and existing unit tests pass locally with my changes
- [ no ] Any dependent changes have been merged and published in downstream modules
- [ yes ] All the sanity checks have been completed and the sanity test cases have been executed

## Ansible Best Practices
- [ yes ] Tasks are idempotent (can be run multiple times without changing state)
- [ no ] Variables and secrets are handled securely (e.g., using `ansible-vault` or environment variables)
- [ yes ] Playbooks are modular and reusable
- [ no ] Handlers are used for actions that need to run on change

## Documentation
- [ yes ] All options and parameters are documented clearly.
- [ yes ] Examples are provided and tested.
- [ yes ] Notes and limitations are clearly stated.

### CODE REVIEW

File: plugins/modules/user_role_workflow_manager.py [hunk 1]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+        Generate a new encryption key using Fernet.
+        Returns:
+            - key (bytes): A newly generated encryption key.
+        Criteria:
+            - This function should only be called if HAS_FERNET is True.
+        """
+        if HAS_FERNET:
+            return Fernet.generate_key()
+        else:
+            self.msg = "'pip install cryptography' library is not installed."
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:999]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py [hunk 2]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+            self.log(self.msg, "ERROR")
+            self.status = "failed"
+            return self
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1002]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py [hunk 3]
```diff
@@ -868,6 +875,7 @@ def __init__(self, module):
         self.created_user, self.updated_user, self.no_update_user = [], [], []
         self.created_role, self.updated_role, self.no_update_role = [], [], []
         self.deleted_user, self.deleted_role = [], []
+        self.key = self.generate_key()
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:878]: I have updated the code and handled this situation in the code by adding the key in validate function

File: plugins/modules/user_role_workflow_manager.py [hunk 4]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+        """
+        Generate a new encryption key using Fernet.
+        Returns:
+            - key (bytes): A newly generated encryption key.
+        Criteria:
+            - This function should only be called if HAS_FERNET is True.
+        """
+        if HAS_FERNET:
+            return Fernet.generate_key()
+        else:
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:998]: I have changed the accordingly to handle the IF statement in validation function itself

        if HAS_FERNET:
            self.key = self.generate_key()
        else:
            self.msg = "The 'cryptography' library is not installed. Please install it using 'pip install cryptography'."
            self.log(self.msg, "ERROR")
            self.status = "failed"
            return self

File: plugins/modules/user_role_workflow_manager.py [hunk 5]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+        Args:
+            - password (str): The plaintext password to be encrypted.
+            - key (bytes): The encryption key used to encrypt the password.
+        Returns:
+            - encrypted_password (bytes): The encrypted password as bytes.
+        Criteria:
+            - This function should only be called if HAS_FERNET is True.
+            - The password should be encoded to bytes before encryption.
+        """
+        if HAS_FERNET:
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1016]: I have changed the accordingly to handle the IF statement in validation function itself

File: plugins/modules/user_role_workflow_manager.py [hunk 6]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+            fernet = Fernet(key)
+            encrypted_password = fernet.encrypt(password.encode())
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1018]: I have changed the accordingly to handle the IF statement in validation function itself

[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1018]: I have changed the accordingly to handle the IF statement in validation function itself

File: plugins/modules/user_role_workflow_manager.py [hunk 7]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+            return encrypted_password
+
+    def decrypt_password(self, encrypted_password, key):
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1021]: I have changed the accordingly to handle the IF statement in validation function itself

File: plugins/modules/user_role_workflow_manager.py [hunk 8]
```diff
@@ -973,6 +985,56 @@ def validate_string_field(self, field_value, regex, error_message, error_message
...
+            - key (bytes): The encryption key used to decrypt the password.
+        Returns:
+            - decrypted_password (str): The decrypted plaintext password.
+        Criteria:
+            - This function should only be called if HAS_FERNET is True.
+            - The encrypted password should be decoded from bytes after decryption.
+        """
+        if HAS_FERNET:
+            fernet = Fernet(key)
+            decrypted_password = fernet.decrypt(encrypted_password.encode()).decode()
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1035]: I have changed the accordingly to handle the IF statement in validation function itself.

I have raised exception that 'cryptography' not installed in 'validate_input_yml' function.

File: plugins/modules/user_role_workflow_manager.py [hunk 9]
```diff
@@ -1179,7 +1241,10 @@ def valid_user_config_parameters(self, user_config):
         password = user_config.get("password")
 
         if password:
-            self.validate_password(password, error_messages)
+            user_config['password'] = self.decrypt_password(password, self.key)
+            plain_password = user_config.get("password")
+            self.validate_password(plain_password, error_messages)
+            user_config['password'] = self.encrypt_password(plain_password, self.key).decode()
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1247]: Updated the code accordingly.

File: plugins/modules/user_role_workflow_manager.py [hunk 10]
```diff
@@ -1504,6 +1569,8 @@ def create_user(self, user_params):
             - Logs the provided user parameters and the received API response.
             - Returns the API response from the "create_user" function.
         """
+        self.log("Create user with user_info_params: {0}".format(str(user_params)), "DEBUG")
```
[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1572]: Ok Madhan will remove the log

[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py:1572]: Will remove the 'user_params' from the log and stating "Create user with user_info_params...."

File: plugins/modules/user_role_workflow_manager.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (madhansansel) [reply plugins/modules/user_role_workflow_manager.py:1022]: Can't we move all these functions in dnac.py module?

[the remark this answers is not in the document]
Comment (ajithandrewj) [reply plugins/modules/user_role_workflow_manager.py]: I have moved the import and functions to the dnac.py file

File: plugins/modules/user_role_workflow_manager.py [hunk 11]
```diff
@@ -926,6 +933,18 @@ def validate_input_yml(self, user_role_details):
             return self
 
         if "user_details" in config and "username" in user_role_details[0] or "email" in user_role_details[0]:
+            for user in user_role_details:
+                if 'password' in user:
+                    encrypt_password_responce = self.encrypt_password(user['password'], self.key.get("generate_key"))
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:938]: encrypt_password_response

File: plugins/modules/user_role_workflow_manager.py [hunk 12]
```diff
@@ -1179,7 +1198,27 @@ def valid_user_config_parameters(self, user_config):
         password = user_config.get("password")
 
         if password:
-            self.validate_password(password, error_messages)
+            decrypt_password_responce = self.decrypt_password(password, self.key.get("generate_key"))
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1201]: decrypt_password_response

File: plugins/modules/user_role_workflow_manager.py [hunk 13]
```diff
@@ -1504,6 +1543,16 @@ def create_user(self, user_params):
             - Logs the provided user parameters and the received API response.
             - Returns the API response from the "create_user" function.
         """
+        self.log("Create user with 'user_params' argument...", "DEBUG")
+        decrypt_password_responce = self.decrypt_password(user_params['password'], self.key.get("generate_key"))
+
+        if decrypt_password_responce and "error_message" in decrypt_password_responce:
```
Comment (madhansansel) [inline plugins/modules/user_role_workflow_manager.py:1549]: response

**Review by madhansansel — APPROVED** [review]

**Pull request merged**